packages/iov-bns/src/bnsconnection.spec.ts

```diff
@@ -1220,6 +1222,132 @@ describe("BnsConnection", () => {
 
       connection.disconnect();
     });
+
+    it("can create and vote on a proposal", async () => {
+      pendingWithoutBnsd();
+      const connection = await BnsConnection.establish(bnsdTendermintUrl);
+      const chainId = connection.chainId();
+
+      const { profile, faucet: author } = await userProfileWithFaucet(chainId);
+      const authorAddress = identityToAddress(author);
+
+      const someElectionRule = (await connection.getElectionRules()).find(
+        // Dictatorship electorate
+        ({ electorateId }) => electorateId === 2,
+      );
+      if (!someElectionRule) {
+        throw new Error("No election rule found");
+      }
+      const startTime = Math.floor(Date.now() / 1000) + 3;
+
+      const title = `Hello ${Math.random()}`;
+      const description = `Hello ${Math.random()}`;
+      const option = `The winner is Alice ${Math.random()}`;
+      let proposalId: string;
+
+      {
+        const createProposal = await connection.withDefaultFee<CreateProposalTx & WithCreator>({
+          kind: "bns/create_proposal",
+          creator: author,
+          title: title,
+          description: description,
+          author: authorAddress,
+          electionRuleId: someElectionRule.id,
+          option: option,
+          startTime: startTime,
+        });
+        const nonce = await connection.getNonce({ pubkey: author.pubkey });
+        const signed = await profile.signTransaction(createProposal, bnsCodec, nonce);
+        const response = await connection.postTx(bnsCodec.bytesToPost(signed));
+        const blockInfo = await response.blockInfo.waitFor(info => !isBlockInfoPending(info));
+        if (!isBlockInfoSucceeded(blockInfo)) {
+          throw new Error("Transaction did not succeed");
+        }
+        if (!blockInfo.result) {
+          throw new Error("Transaction result missing");
+        }
+        proposalId = toHex(blockInfo.result).toUpperCase();
+      }
+
+      {
+        // Election submitted, voting period not yet started
+        const proposal = (await connection.getProposals()).find(p => p.id === proposalId)!;
+        expect(proposal.votingStartTime).toBeGreaterThan(Date.now() / 1000);
```

Comment:
The timing conditions are less about how the code works and more about how the test setup works. If the transaction fails we throw an error rather than failing an expectation. Should we also do that for these timing-related requirements?

Reply:
A proper [testing framework](https://stackoverflow.com/a/2565309/2013738) comes with [built-in expectations and assetions](https://github.com/catchorg/Catch2/blob/master/docs/assertions.md#natural-expressions), where assertions stop test execution and expectations don't. Unfortunately, everything in the JS world does not have that, which causes a bit of strangenes in code like:

```js
expect(encoded.sendMsg).toBeDefined(); // test should fail and end here
expect(encoded.sendMsg!.memo).toEqual("free transaction"); // but it only ends when accessing `undefined.memo`, creating a reporting that is misleading
```

What we have is not great but does the job:

* `expect` is the most readable option in most cases, especially before we allowed `if`s without curly brackets
* `throw` stops execution, which is nice to have. Most importantly, it is used by the TypeScript type analyzer to infer correct types in the non-error case.

So the main reason for using exception for transaction fails is that otherwise `blockInfo.result` would not be available below. My rule of thumb so far is: use `expect` by default for readability and `throw` when TS needs it.

Reply:
How would you feel about using [Node assert](https://nodejs.org/api/assert.html#assert_assert) for these assertions which should stop execution if they fail?

Reply:
Would be good to have something like that. Unfortunately, Node assert does not play well with TypeScript, as TS does not get the information that for some cases an exception is thrown. So

```
assert(encoded.sendMsg);
expect(encoded.sendMsg.memo).toEqual("free transaction"); // encoded.sendMsg is still cash.ISendMsg | null | undefined at compile time
```

As far as I know there is no way to express _function will throw type X if runtime condition y fails_ using TypeScript declarations. If there was, we could build a small assertion library on our own.